logicalplan/plan.go

```diff
@@ -16,33 +17,66 @@ var DefaultOptimizers = []Optimizer{
 }
 
 type Plan interface {
-	Optimize([]Optimizer) Plan
+	Optimize(...Optimizer)
+	OptimizationsApplied() []string
+
+	PreOptimizationExpr() parser.Expr
 	Expr() parser.Expr
 }
 
 type Optimizer interface {
-	Optimize(parser.Expr) parser.Expr
+	Optimize(parser.Expr, *Log) parser.Expr
 }
 
 type plan struct {
-	expr parser.Expr
+	initial string
+	expr    parser.Expr
+	optLog  *Log
 }
 
 func New(expr parser.Expr, mint, maxt time.Time) Plan {
 	expr = promql.PreprocessExpr(expr, mint, maxt)
 	setOffsetForAtModifier(mint.UnixMilli(), expr)
 
 	return &plan{
-		expr: expr,
+		initial: expr.String(),
+		expr:    expr,
+		optLog:  &Log{},
+	}
+}
+
+func (p *plan) OptimizationsApplied() []string {
+	ret := make([]string, 0, len(p.optLog.Elems()))
+	for _, e := range p.optLog.Elems() {
+		ret = append(ret, fmt.Sprintf("Logical Optimization -> %v", e))
+	}
+	return ret
+}
+
+type Log struct {
```

Comment:
This is great way of recording optimizations.  We can likely use the same when we have physical phase optimizations!